### PR TITLE
Add standard fields property and fetchAll method to ProductFournisseur class

### PR DESCRIPTION
**Pull request by AlexisLaurier**
# Add fields property to ProductFournisseur class in order to be able to use standard methods from common object + fetchAll method

### CODE REVIEW

File: htdocs/fourn/class/fournisseur.product.class.php [hunk 1]
```diff
@@ -134,7 +134,43 @@ class ProductFournisseur extends Product
 
 	public $packaging;
 
-
+	public $fields=array(
```
Comment (eldy) [inline htdocs/fourn/class/fournisseur.product.class.php:137]: The properties of class ProductFournisseur should be same than properties of class Product + new ones (the class extedns Product).
So i suggest to not declare the ->fields here (it is already declared into product) and just add into the constructor, the new fileds

$this->fields = array_merge($this->fields, array of new fields)

So the ProductFournisseur will avoid duplicate declaration and ProductFournisseur will always be upt to date compared to Product.

Comment (AlexisLaurier) [reply htdocs/fourn/class/fournisseur.product.class.php]: Ok that what I did but I had to overwrite the getFieldList method as this method reflects the database fields.

File: htdocs/fourn/class/fournisseur.product.class.php [hunk 2]
```diff
@@ -147,6 +146,37 @@ public function __construct($db)
...
+			'remise' => array('type'=>'double', 'label'=>'Remise', 'enabled'=>'1', 'position'=>70, 'notnull'=>1, 'visible'=>-1,),
+			'unitprice' => array('type'=>'double(24,8)', 'label'=>'Unitprice', 'enabled'=>'1', 'position'=>75, 'notnull'=>0, 'visible'=>-1,),
+			'charges' => array('type'=>'double(24,8)', 'label'=>'Charges', 'enabled'=>'1', 'position'=>80, 'notnull'=>0, 'visible'=>-1,),
+			'default_vat_code' => array('type'=>'varchar(10)', 'label'=>'Defaultvatcode', 'enabled'=>'1', 'position'=>85, 'notnull'=>0, 'visible'=>-1,),
+			'tva_tx' => array('type'=>'double(6,3)', 'label'=>'Tvatx', 'enabled'=>'1', 'position'=>90, 'notnull'=>1, 'visible'=>-1,),
+			'info_bits' => array('type'=>'integer', 'label'=>'Infobits', 'enabled'=>'1', 'position'=>95, 'notnull'=>1, 'visible'=>-1,),
+			'fk_user' => array('type'=>'integer:User:user/class/user.class.php', 'label'=>'Fkuser', 'enabled'=>'1', 'position'=>100, 'notnull'=>0, 'visible'=>-1,),
+			'fk_supplier_price_expression' => array('type'=>'integer', 'label'=>'Fksupplierpriceexpression', 'enabled'=>'1', 'position'=>105, 'notnull'=>0, 'visible'=>-1,),
```
Comment (eldy) [inline htdocs/fourn/class/fournisseur.product.class.php:165]: This fields does not seems to be field of the class ProductFournisseur (related to table llx_product) but are related to table llx_product_fournisseur_price (so the table of Vendor Prices).
Such fields must be into the class than manage ProductFournisseurPrice and not int the class ProductFournisseur. If such class/files does not exists, it must be introduced as a new class/file. 1 table = 1 class.

File: htdocs/fourn/class/fournisseur.product.class.php [hunk 3]
```diff
@@ -1241,4 +1271,96 @@ private function logPrice($user, $datec, $buyprice, $qty, $multicurrency_buypric
...
+     *
+     * @param  string      $sortorder    Sort Order
+     * @param  string      $sortfield    Sort field
+     * @param  int         $limit        limit
+     * @param  int         $offset       Offset
+     * @param  array       $filter       Filter array. Example array('field'=>'valueforlike', 'customurl'=>...)
+     * @param  string      $filtermode   Filter mode (AND or OR)
+     * @return array|int                 int <0 if KO, array of pages if OK
+     */
+    public function fetchAll($sortorder = '', $sortfield = '', $limit = 0, $offset = 0, array $filter = array(), $filtermode = 'AND')
```
Comment (eldy) [inline htdocs/fourn/class/fournisseur.product.class.php:1286]: This method does not seems to be field of the class ProductFournisseur (related to table llx_product) but are related to table llx_product_fournisseur_price (so the table of Vendor Prices).
So this method must be available into a new class/file ProductFournisseurPrice

Comment (AlexisLaurier) [reply htdocs/fourn/class/fournisseur.product.class.php:1286]: @eldy Where do you prefer to put this CRUD class ? into htdocs/fourn/class or htdocs/product/class ?

Comment (eldy) [reply htdocs/fourn/class/fournisseur.product.class.php:1286]: Good question.
Let's say into htdocs/product/class
I have a hope to see one day the fourn directory be removed and replaced into supplier_order and supplier_invoice.

**Comment by AlexisLaurier** [comment]
@eldy i've created the new productfournisseurprice class and reverted my previous modifications

**Pull request merged**